gh-55454: Add IMAP4 IDLE support to imaplib (#122542)
python/cpython: Doc/library/imaplib.rst

.. changes for IMAP4_SSL by Tino Lange <Tino.Lange@isg.de>, March 2002
.. changes for IMAP4_stream by Piers Lauder <piers@communitysolutions.com.au>,
   November 2002

> .. changes for IDLE by Forest <forestix@nom.one> August 2024

Reviewer: ```suggestion
.. changes for IMAP4 IDLE by Forest <forestix@nom.one> August 2024
```
(Avoid using just IDLE anywhere, since IDLE is also Pythons stdlib editing app -- I feel it's important to clarify this.)

> .. changes for IMAP4 IDLE by Forest <forestix@nom.one> August 2024

Reviewer 2: ```suggestion
.. changes for IMAP4 IDLE by Forest <forestix@nom.one>, August 2024
```

Author: Fixed in 3221dbc0e302.

   of the IMAP4 QUOTA extension defined in rfc2087.

> .. method:: IMAP4.idle(duration=None)

Reviewer 2: I missed this one in my first round of review, but you might want to use `.. versionadded:: next` somewhere to indicate when this feature has been added (i.e. which version) (The `next` would automatically be converted into 3.14).

I don't really know whether it's better to place it before or after the example. A similar argument applies to `burst`. I'll let you decide for that.

Author: Added in 40e607a209d.

I used a single `versionadded`. I think that's sufficient because there has never been a version of `Idler` without `Idler.burst()`, and since both require a call to `idle()` (which now has a `versionadded`).

I placed it at the bottom of the section because:

- It applies to everything in that section.
- It fits will with the notes that are already at the bottom.
- When I tried it just before/after the example code, I found that it distracted from reading and understanding how the API works.

Reviewer 2: Thanks for that. If we ever need to add additional parameters to either of the two methods, we'll duplicate the `versionadded` and add a `versionchanged` so that it's similar to the rest of the docs. But for now it's perfect (and I don't think we'll ever change this interface, unless for some good reason)

   Return an :class:`!Idler`: an iterable context manager implementing the
   IMAP4 ``IDLE`` command as defined in :rfc:`2177`.

   The returned object sends the ``IDLE`` command when activated by the
   :keyword:`with` statement, produces IMAP untagged responses via the
   :term:`iterator` protocol, and sends ``DONE`` upon context exit.

>    Responses are represented as ``(type, [data, ...])`` tuples, as described

Reviewer 3: Still think the documentation could explain where responses end up that are sent before _idle_ starts. I don’t think they would be returned by IMAP4.response.

```suggestion
   The :class:`!Idler` yields any untagged responses sent before the
   command started, as well as responses to the ``IDLE`` command.
   Responses are represented as ``(type, [data, ...])`` tuples, as described
```

Author: > I don’t think they would be returned by IMAP4.response.

They would be after IDLE ends, if they were not retrieved using the iterator.

The first part of that suggestion gives me the impression that the iterator will produce untagged responses that were collected before the IDLE command is sent. I think you're referring to untagged responses arriving after we send IDLE but before the continuation response arrives, yes?

The second part is incorrect, since responses to the IDLE command are not untagged responses at all, and will never be produced by the iterator. I think you're referring to untagged responses arriving after we send DONE but before the idle state ends, yes?

How about this:

> All untagged responses arriving after the `IDLE` command is sent (including any that arrive before the server acknowledges the command) will be available via iteration. Any responses that are not iterated in the *with* context can be retrieved afterward with `IMAP4.response()`.

Author: Revised suggestion:

> All untagged responses arriving after the `IDLE` command is sent (including any that arrive before the server acknowledges the command) will be available via iteration. Any leftover responses (those not iterated in the *with* context) can be retrieved in the usual way after IDLE ends, using `IMAP4.response()`.

Author: Done in e8a85094b2a.

      next response along with any immediately available subsequent responses.
      (For example, a rapid series of ``EXPUNGE`` responses after a bulk
      delete.)

      Requires a socket connection; does not work on :class:`IMAP4_stream`
      connections.

      .. code-block:: pycon

         >>> with M.idle() as idler:
         ...     # get a response and any others following by < 0.1 seconds
         ...     batch = list(idler.burst())
         ...     print(f'processing {len(batch)} responses...')
         ...     print(batch)
         ...
         processing 3 responses...
         [('EXPUNGE', [b'2']), ('EXPUNGE', [b'1']), ('RECENT', [b'0'])]

      .. tip::

         The ``IDLE`` context's maximum duration, as passed to
         :meth:`IMAP4.idle`, is respected when waiting for the first response
         in a burst. Therefore, an expired :class:`!Idler` will cause this
         generator to return immediately without producing anything. Callers
         should consider this if using it in a loop.

   .. note::

      The :class:`!Idler` object returned by :meth:`IMAP4.idle` is usable only
      within a :keyword:`with` statement. To retrieve unsolicited IMAP

>       responses outside that context, see :meth:`IMAP4.response`.

Reviewer 3: ```suggestion
      responses after the context exited, see :meth:`IMAP4.response`.
```

Author: This applies not only after exiting the context, but also before entering it, and even when no such context has been created.

How about this:

"To retrieve unsolicited IMAP responses before entering or after leaving that context,"

Reviewer 3: My problem is it gives the impression you can poll the _response_ method (like you can poll the iterator) to get responses as they are sent from the server. Whaddabout something like

> The idler object is usable only within a _with_ statement. _IMAP4.response_ must used before entering that context to retrieve unsolicited responses collected by the previous command, and after leaving the context to retrieve any remaining responses received before the IDLE command terminated.

Author: That phrasing gives me the impression that I must call `response()` before calling `idle()`, which is not true.

How about this:

> The iterator returned by `IMAP4.idle()` is usable only within the scope of a *with* statement. Before or after that context, unsolicited responses are collected internally whenever a command finishes, and can be retrieved with `IMAP4.response()`.

Author: Done in 8d7801054b7.

Marking resolved for now. We can always re-open if there's a problem with it.

   of the IMAP4 QUOTA extension defined in rfc2087.

.. method:: IMAP4.idle(dur=None)

   Return an iterable context manager implementing the ``IDLE`` command
   as defined in :rfc:`2177`.

   The context manager sends the ``IDLE`` command when activated by the
   :keyword:`with` statement, produces IMAP untagged responses via the
   :term:`iterator` protocol, and sends ``DONE`` upon context exit.

   The *dur* argument sets a maximum duration (in seconds) to keep idling,
   after which iteration will stop.  It defaults to ``None``, meaning no time
   limit.  Callers wishing to avoid inactivity timeouts on servers that impose

>    them should keep this <= 29 minutes.

Reviewer 4: I'm not a fan of using operators in text. How about this?
```suggestion
   them should keep this at most 29 minutes.
```

   See the :ref:`warning below <windows-pipe-timeout-warning>` if using
   :class:`IMAP4_stream` on Windows.

   Response tuples produced by the iterator almost exactly match those

>    returned by other imaplib methods.  The difference is that the tuple's

Reviewer 4: This *is* imaplib, why refer to it in the third person?
```suggestion
   returned by other methods in this module.  The difference is that the tuple's
```

Author: Referring to imaplib in the third person became something of a habit after a couple of months developing and testing the code in a separate module. :) Good point. I'll update it.

   second member is a single response datum, rather than a list of data.
   Therefore, in a mailbox where calling ``M.response('EXISTS')`` would
   return ``('EXISTS', [b'1'])``, the idle iterator would produce
   ``('EXISTS', b'1')``.

   Example::

      with M.idle(dur=29*60) as idler:
          for response in idler:
              typ, datum = response

>               print(typ, datum)

Reviewer 4: I think we can format this example a little better. Namely, there should be a space between operators (per PEP 8), and the `typ, datum = response` line is extra work -- the loop can just get refactored to `for typ, datum in idler`
```suggestion
      with M.idle(dur=29 * 60) as idler:
          for type, datum in idler:
              print(typ, datum)
```

Author: The extra line was there with a purpose: It demonstrated the relationship between a "response" described in the text and the `(type, datum)` data structure familiar to readers who have used imaplib before.

I used the same example when [first describing my implementation](https://github.com/python/cpython/issues/55454#issuecomment-2227543041) in the bug report, where it also served to highlight a difference between my approach and others that had been proposed.

Yes, it is an extra step that could be factored out, and I trusted the reader to know python well enough to see that immediately.

On the other hand, I also value simplicity, so factoring it out of the example makes some sense as well. I'm fine with it either way.

Reviewer 4: Many will copy-paste from the documentation and won't bother to refactor it. IMO, it's pretty clear that there's a relationship; an extra `response` variable doesn't add much.

      ('EXISTS', b'1')
      ('RECENT', b'1')

   Instead of iterating one response at a time, it is also possible to retrieve
   the next response along with any immediately available subsequent responses
   (e.g. a rapid series of ``EXPUNGE`` events from a bulk delete).  This
   batch processing aid is provided by the context's ``burst()``
   :term:`generator`:

>    .. method:: idler.burst(interval=0.1)

Reviewer 4: I think we should rename `_Idler` to `Idler`, because prefixing it with `_` forces compromises like this to exist. `idler` is not something that imaplib has, so this is a little ambiguous. It should be `Idler.burst`, and the `idle` method should document that it returns an `Idler`.

Author: > prefixing it with _ forces compromises like this to exist.

Noticing that the name `idler` was a deliberate compromise is an astute observation. :)

You convinced me (in an earlier suggestion) to rename `_Idler` to `Idler` and give it a doc string. I'll make all the related changes in one commit.

Author: Done in 94c02e88a9d3.

      Yield a burst of responses no more than *interval* seconds apart.

      Example::

         with M.idle() as idler:

             # get the next response and any others following by < 0.1 seconds
             batch = list(idler.burst())

             print(f'processing {len(batch)} responses...')
             print(batch)

         processing 3 responses...
         [('EXPUNGE', b'2'), ('EXPUNGE', b'1'), ('RECENT', b'0')]

      The ``IDLE`` context's maximum duration (the *dur* argument to
      :meth:`IMAP4.idle`) is respected when waiting for the first response

>       in a burst.  Therefore, an expired idle context will cause this generator

Reviewer 4: Let's stay consistent with the use of "idle"
```suggestion
      in a burst.  Therefore, an expired ``IDLE`` context will cause this generator
```

Author: Referring informally to the idle mode vs. referring formally to the IDLE command. Fair enough. :)

      to return immediately without producing anything.  Callers should
      consider this if using it in a loop.

   .. _windows-pipe-timeout-warning:

   .. warning::

>       Windows :class:`IMAP4_stream` connections have no way to accurately
>       respect the *dur* or *interval* arguments, since Windows ``select()``
>       only works on sockets.

Reviewer 4: This is something that Windows *has*, so it needs to be possessive.
```suggestion
      Windows' :class:`IMAP4_stream` connections have no way to accurately
      respect the *dur* or *interval* arguments, since Windows' ``select()``
      only works on sockets.
```

   of the IMAP4 QUOTA extension defined in rfc2087.

.. method:: IMAP4.idle(dur=None)

>    Return an ``Idler``: an iterable context manager implementing the ``IDLE``

Reviewer 4: ```suggestion
   Return an :class:`Idler`: an iterable context manager implementing the ``IDLE``
```

>    Return an :class:`!Idler`: an iterable context manager implementing the ``IDLE``

Reviewer: ```suggestion
   Return an :class:`!Idler`: an iterable context manager implementing the IMAP4 ``IDLE``
```

   command as defined in :rfc:`2177`.

   The context manager sends the ``IDLE`` command when activated by the

>    :keyword:`with` statement, produces IMAP untagged responses via the

Reviewer 4: ```suggestion
   :keyword:`with` statement, which produces IMAP untagged responses via the
```

Author: This change would make the text incorrect, since it is the iterator that produces untagged responses, not the `with` statement.

Counterproposal: I could change "The context manager sends" to "The returned object sends". This could perhaps make the intent more clear in light of the fact that the object is both a context manager and an iterator.

>    :term:`iterator` protocol, and sends ``DONE`` upon context exit.

Reviewer 3: Is it worth clarifying that the Idler iterable should only be iterated inside its _with_ statements, i.e. you have to enter _with_ to get unsolicited responses sent before the _idle_ command started, and you have to use the general _response_ method after exit

Author: Added as a note  in 92d7ce7ca4ed.

   The *dur* argument sets a maximum duration (in seconds) to keep idling,

>    after which iteration will stop.  It defaults to ``None``, meaning no time

Reviewer 4: ```suggestion
   after which the iteration will stop.  It defaults to ``None``, meaning no time
```

Author: This change would alter the meaning of the sentence. As currently written, it means that no more iterations will occur. With the proposed change, it would mean that the current iteration stops, but would say nothing about subsequent iterations.

Reviewer 4: How about "the underlying iteration"?

Author: How about "any ongoing iteration will stop"? I think that's more like what you originally suggested, and it avoids referring to an "underlying iteration" that might not exist (for example, if it hasn't started yet).

   meaning no time limit.  Callers wishing to avoid inactivity timeouts on

>    servers that impose them should keep this at most 29 minutes.

Reviewer 2: I would also mention that 29 minutes is 1740 seconds since the duration's unit is the second.

```suggestion
   servers that impose them should keep this at most 29 minutes (1740 seconds).
```

Author: Done in 171ebf1b65bb.

   See the :ref:`warning below <windows-pipe-timeout-warning>` if using
   :class:`IMAP4_stream` on Windows.

>    Response tuples produced by the iterator almost exactly match those

Reviewer 2: I'm not a native English speaker but I would simply say "almost match" because having "almost exactly" rings contradictory.

Author: Thanks for mentioning this. The phrase "almost exactly" is widely accepted (without a hint of contradiction) where I live, but non-native speakers also matter, and connotations sometimes vary even among English-speaking regions.

This change is no longer necessary, though, since the tuples now *exactly* match those returned by other methods. The sentence was removed in 9c4af2c4e7e0.

   returned by other methods in this module.  The difference is that the tuple's
   second member is a single response datum, rather than a list of data.
   Therefore, in a mailbox where calling ``M.response('EXISTS')`` would
   return ``('EXISTS', [b'1'])``, the idle iterator would produce
   ``('EXISTS', b'1')``.

   Example::

      with M.idle(dur=29 * 60) as idler:
          for typ, datum in idler:
              print(typ, datum)

      ('EXISTS', b'1')

>       ('RECENT', b'1')

Reviewer 3: Is this a mistake? Why is this apparently printing a single tuple each loop? It reminds me of trying to call the Python 2 _print_ statement.

Might also be worth using the >>> prompt notation like done under the IMAP4 constructor.

Reviewer 2: I think it's better to use a `pycon` code-block together with the `>>>` notation. Here, the exemple would render as a single code block and `('EXIST', b'1')` would be a statement (that does nothing).

   Instead of iterating one response at a time, it is also possible to retrieve
   the next response along with any immediately available subsequent responses

>    (e.g. a rapid series of ``EXPUNGE`` events from a bulk delete).  This

Reviewer 2: ```suggestion
   (e.g., a rapid series of ``EXPUNGE`` events from a bulk delete).  This
```

Author: Rephrased as "For example," in b73a3656c4b0.

>    batch processing aid is provided by the context's ``burst()``

Reviewer 4: Can we make these references?
```suggestion
   batch processing aid is provided by the context's :meth:`imaplib.Idler.burst`
```

Author: The `Idler` class is deliberately undocumented, and therefore not a valid link target. [Remember?](https://github.com/python/cpython/pull/122542#discussion_r1776026417) :)

Author: I'll see if I can get reST to link directly to the method definition anyway.

Reviewer 4: I think we can still link to methods.

Author: Yes, it looks like I can do that. (At least, it works in my local docs build.)

It does invite the chance of a reader mistaking two adjacent links for a single one, but I suppose that's not exactly unheard of in hypertext, and the underlined link style used by docs.python.org should mitigate that risk.

It also links to the text immediately below itself, which might be considered silly, but I suppose it makes some sense for consistency given that we're making the same change later in the document.

Okay, will do.

Author: This is what I was afraid of. It generates working links in the HTML, but causes reST warnings that break the continuous integration tests:

```
Run python Doc/tools/check-warnings.py \
Emitting doc warnings matching modified lines:
Warning: py:meth reference target not found: Idler.burst [ref.meth]
/home/runner/work/cpython/cpython/Doc/library/imaplib.rst:389: WARNING: py:meth reference target not found: Idler.burst [ref.meth]
Warning: py:meth reference target not found: Idler.burst [ref.meth]
/home/runner/work/cpython/cpython/Doc/library/imaplib.rst:389: WARNING: py:meth reference target not found: Idler.burst [ref.meth]
Error: must not contain warnings:
Doc/library/imaplib.rst
  389: py:meth reference target not found: Idler.burst [ref.meth]
  389: py:meth reference target not found: Idler.burst [ref.meth]
Error: Process completed with exit code 255.
```

Author: Last time we ran into this, we solved it with [cpython's `!` reST extension](https://devguide.python.org/documentation/markup/#roles). That disables links, though, which I think was what you were after when suggesting the change.

Also, I just noticed that the link generation only works in the first instance. The later two, which occur in the "warning" section, weren't converted to links. Perhaps the first one was just a fluke, and shouldn't have worked in the first place given that it referenced an undefined class?

I considered inserting a reST label where the method is defined, and linking to that instead, but [the cpython dev guide says](https://devguide.python.org/documentation/markup/#cross-linking-markup) this requires also inserting a section title, which I think would make the docs harder to read rather than easier.

Unless you have a good solution for this, I'm inclined to revert it.

Author: I also tried a plain internal hyperlink, but that stripped away the monospaced font, making the method name look like it was trying to be a regular word. (It also made the reST source ugly.)

Author: I've reverted this because it breaks CI (and doesn't work properly anyway). If it's important to you, I am open to suggestions on how to make it work.

For now, I've returned it to inline literal syntax, which I think makes the meaning clear, especially since this text is in close proximity to the method's signature and example code. I also rephrased `burst()` to `Idler.burst()`, for an abundance of clarity.

Do you think that's good enough to move the PR forward?

Reviewer 3: Maybe the problem with linking is because in the documentation, the Idler class seems to be defined inside the IMAP4 class. Try linking to imaplib.IMAP4.Idler.burst, or move the Idler class outside of the IMAP class? (In the implementation it is at module level: imaplib.Idler.)

Reviewer 4: Sorry for not responding quickly. I'd like some sort of reference, regardless of whether its hyperlinked or not; `burst()` without context of the class can be confusing for users.

Author: [vadmium](https://github.com/vadmium) wrote:

> Maybe the problem with linking is because in the documentation, the Idler class seems to be defined inside the IMAP4 class. Try linking to imaplib.IMAP4.Idler.burst

Ah, yes, that does get sphinx to make a working link. However, it also burdens the reader with a bunch of unimportant namespacing, and cements where `Idler` is defined (which we do not want, for the same reasons that it is a deliberately undocumented class).

> or move the Idler class outside of the IMAP class? 

It originally was that way, but it made the documentation needlessly obtuse, requiring the reader to jump from one section to another (and back again) to understand how to use `idle()`. Moving it to where it is now made learning how to use it noticeably easier, and since `Idler` is only ever used with `idle()` and has only one public method, I think it's fine to nest its documentation to present it in context.

[ZeroIntensity](https://github.com/ZeroIntensity) wrote:

> I'd like some sort of reference, regardless of whether its hyperlinked or not; burst() without context of the class can be confusing for users.

Yes, that makes sense. Note that the sentence where it is first used provides the context by saying, "the context object's `burst()` generator". And when I reverted the broken-link change, I also added context by changing `burst()` to `Idler.burst()`. I suppose I could make the same change in that first sentence, so it would read, "the context object's `Idler.burst()` generator". Would you prefer that?

Reviewer 2: > Ah, yes, that does get sphinx to make a working link. However, it also burdens the reader with a bunch of unimportant namespacing, and cements where Idler is defined (which we do not want, for the same reasons that it is a deliberately undocumented class).

You can change how the link is being rendered as follows:

```rst
:meth:`Idler.burst <imaplib.IMAP4.Idler.burst>
```

This will render `Idler.burst` but will use the correct link.

Author: Thanks. It makes the links large and ugly in reST...

    :meth:`Idler.burst() <imaplib.IMAP4.Idler.burst>`

...but I think that's fine, since few people read the reST source.

Done in 5d8a40be0cb.

>    :term:`generator`:

Reviewer 3: I wonder if it would be better to move this paragraph under the Idler.burst heading? I missed the clue that _burst_ is a generator the first few passes through trying to understand the changes.

Author: Done (and rephrased accordingly) in b73a3656c4b0.

>    :term:`generator`:

Reviewer 2: ```suggestion
   :term:`generator`.
```

I understand why you would use a colon but we generally consider `.. method::` blocks to be standalone paragraphs.

Author: Fixed in 5d8a40be0cb.

   .. method:: Idler.burst(interval=0.1)

      Yield a burst of responses no more than *interval* seconds apart.

      Example::

>          with M.idle() as idler:

Reviewer 2: Ditto for using `pycon`

Author: These examples were replaced in 9c4af2c4e7e0 and now use `>>>` notation, as you suggested earlier.

If you still want the `pycon` directive added, can you please show me where it is documented? I didn't see it in the reStructuredText docs, nor in the [Python Developer's Guide](https://devguide.python.org/documentation/markup/). The stdlib docs seem to use it only in few places, inconsistently, and imaplib is not among them.

Reviewer 2: Ah sorry. Actually, it would be used as follows:

```rst

.. code-block:: pycon

   >>> print(2)
   2
```

This a syntax highlighting provided by Sphinx (based on reST). The docs are more than just reST documents, they are Sphinx-compatible. If you use the `.. code-block::` or `::` markers, you would have a default Python highlighting (as if you were writing a script) but with `pycon`, you would get the same highlighting as if you were writing stuff in the REPL.

I don't remember whether we automatically detect `>>>` for syntax highlighting or not, but it never hurts to be explicit.

Author: Done in 08a45368310e.

FWIW, I didn't see any syntax highlighting difference in my local build of the docs.

             # get the next response and any others following by < 0.1 seconds
             batch = list(idler.burst())

             print(f'processing {len(batch)} responses...')
             print(batch)

         processing 3 responses...
         [('EXPUNGE', b'2'), ('EXPUNGE', b'1'), ('RECENT', b'0')]

>       The ``IDLE`` context's maximum duration (the argument to

Reviewer 2: How about having it as a `.. note::` or a `.. tip::`?

Author: Done in 5682ef4b6c2b.

      :meth:`IMAP4.idle`) is respected when waiting for the first response in a
      burst.  Therefore, an expired ``IDLE`` context will cause this generator
      to return immediately without producing anything.  Callers should
      consider this if using it in a loop.

   .. _windows-pipe-timeout-warning:

   .. warning::

>       Windows' :class:`IMAP4_stream` connections have no way to accurately

Reviewer 2: ```suggestion
      On Windows, :class:`IMAP4_stream` connections have no way to accurately
```

Author: Done in 656e9f5dcd25.

      respect the *dur* or *interval* arguments, since Windows' ``select()``
      only works on sockets.

      If the server regularly sends status messages during ``IDLE``, they will
      wake our iterator anyway, allowing *dur* to behave roughly as intended,
      although usually late.  Dovecot's ``imap_idle_notify_interval`` default
      setting does this every 2 minutes.  Assuming that's typical of IMAP
      servers, subtracting it from the 29 minutes needed to avoid server
      inactivity timeouts would make 27 minutes a sensible value for *dur* in
      this situation.

      There is no such fallback for ``burst()``, which will yield endless
      responses and block indefinitely for each one.  It is therefore advised

>       not to use ``burst()`` with an :class:`IMAP4_stream` connection on

Reviewer 4: ```suggestion
      There is no such fallback for :meth:`imaplib.Idler.burst`, which will yield endless
      responses and block indefinitely for each one.  It is therefore advised
      not to use :meth:`imaplib.Idler.burst` with an :class:`IMAP4_stream` connection on
```

      Windows.

   .. note::

>       Note: The ``Idler`` class name and structure are internal interfaces,

Reviewer 4: ```suggestion
      Note: The :class:`Idler` class name and structure are internal interfaces,
```

Author: I avoided this because I didn't want to create a dangling reference to a class that is deliberately undocumented in the reST file. Since you suggested it, I tried it anyway (locally) and found that the generated html rendered fine, so I committed it.

Unfortunately, it broke the continuous integration checks, which fail with a completely useless error message. (An empty one.) Sigh. With some time and guesswork, I was able to excavate the failing check and run it locally. It confirmed what I originally expected: Dangling references are no good.

    Emitting doc warnings matching modified lines:
    ::warning file=/tmp/cpython/Doc/library/imaplib.rst,line=313::py:class reference target not found: Idler [ref.class]
    /tmp/cpython/Doc/library/imaplib.rst:313: WARNING: py:class reference target not found: Idler [ref.class]
    ::warning file=/tmp/cpython/Doc/library/imaplib.rst,line=396::py:class reference target not found: Idler [ref.class]
    /tmp/cpython/Doc/library/imaplib.rst:396: WARNING: py:class reference target not found: Idler [ref.class]

    Error: must not contain warnings:

    Doc/library/imaplib.rst
      313: py:class reference target not found: Idler [ref.class]
      396: py:class reference target not found: Idler [ref.class]

I've reverted it, at least for now.

I understand the value of reST roles, and have used them throughout the IDLE docs, but in this particular case, I feel the minor benefit is outweighed by the disadvantages:

- It would force adding `Idler` class documentation, thereby promoting a deliberately undocumented internal interface into a documented one, not merely in the doc strings (to support the `help()` builtin), but front-and-center in the module's online docs. In that position, I think it would likely mislead some users into assuming it's okay to use directly, despite the warning against doing so. (Humans are pretty famous for overlooking such warnings.)
- It would mean restructuring the `IDLE` docs in a way that complicates reading and impedes understanding, rather than helping it: Classes are conventionally documented at the top level of the file, so `Idler` would presumably follow suit, and take its public methods with it. This would place `Idler.burst()` way down at the bottom of the page, far apart from `IMAP4.idle()`. Users would no longer be able to read these closely related sections on one screen, but instead have to jump through hoops (following links back and forth between sections) to get the whole picture.
- It would add unnecessary bulk to the docs, burdening users with even more information to take in... for the sake of an internal interface that they shouldn't concern themselves with in the first place. (This sort of thing is why a lot of people complain about man pages. They are put off by being expected to climb a wall of text with substantial parts that don't concern them.)
- It would create a link target, opening the door to other parts of the docs linking to something that they probably shouldn't.

Would this really be worth the drawbacks?

Reviewer 4: In that case, it's an easy fix :)

```suggestion
      Note: The :class:`!Idler` class name and structure are internal interfaces,
```

Author: Aha! I learned something new today. This is apparently a special reST extension just for CPython.

Done in c8d4d6d994e.

>       Note: The :class:`!Idler` class name and structure are internal interfaces,

Reviewer 4: `Note:` is repeated (it's automatically added in the generated documentation)
```suggestion
      The :class:`!Idler` class name and structure are internal interfaces,
```

Author: Oops. I didn't notice when I added that `note::` tag that reST added an extra "Note:" prefix to the text, making the original one redundant. Nice catch.

   .. versionadded:: 3.5

.. _idle context manager:

IDLE Context Manager
--------------------

The object returned by :meth:`IMAP4.idle` implements the context management
protocol for the :keyword:`with` statement, and the :term:`iterator` protocol
for retrieving untagged responses while the context is active.
It also has the following method:

> .. method:: IdleContextManager.burst([interval])

Reviewer 5: Would it be possible to elaborate here (or even show examples) of what `data` should look like? Is it just a string (`EXPUNGE`), or something more complicated? What? How does the arrival of a new message look like?

Author: The data produced by the iterator exactly match those produced by the other imaplib methods, as documented in the existing module's [IMAP4 Objects](https://cpython-previews--122542.org.readthedocs.build/en/122542/library/imaplib.html#imap4-objects) section. (In fact, they originate from the same place.) The only difference is how they are delivered: they are not wrapped in a list as is done by the other methods, since only one is produced with each iteration.

This is mentioned in [the third paragraph](https://cpython-previews--122542.org.readthedocs.build/en/122542/library/imaplib.html#imaplib.IMAP4.idle) of my `idle()` description, and the `burst()` method description states that it produces the same, but since you missed it, I suppose making it more explicit could be worthwhile. I can add a sentence with a small example showing that difference.

I haven't shown the types/formats of the produced values because they are not new with this patch or unique to `idle()`. Rather, they are part of the module's long-standing API. Improving the documentation for them would probably be nice, but also more appropriate in a top-level part of the docs and added by a separate patch, don't you think?

Author: I was a little concerned about cluttering the docs with lengthy details when the surrounding docs are pretty succinct, but if you think it would be worthwhile specifically for IDLE, I could add some typical output to my code examples.

Reviewer 5: > The data produced by the iterator exactly match those produced by the other imaplib methods.

This sentence wouldn’t be a clutter, even if you repeat it couple of times (like here).

And yes, examples are the coolest: am I the only one who starts reading any documentation by looking at examples?

Author: Heh... Good point. I often do find the "show me, don't tell me" approach to be helpful.

Thanks for the feedback. I ended up going further than just adding a sentence, and I'm happier with the result. Please have a look at the docs update that I just pushed.
